controllers/pxc/perconaxtradbcluster_controller.go

```diff
@@ -0,0 +1,250 @@
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	pxcv1 "github.com/Percona-Lab/k8s-vault-issuer/apis/pxc/v1"
+	api "github.com/hashicorp/vault/api"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// PerconaXtraDBClusterReconciler reconciles a PerconaXtraDBCluster object
+type PerconaXtraDBClusterReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=pxc.percona.com,resources=perconaxtradbclusters,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=pxc.percona.com,resources=perconaxtradbclusters/status,verbs=get;update;patch
+
+func (r *PerconaXtraDBClusterReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("perconaxtradbcluster", req.NamespacedName)
+
+	rr := reconcile.Result{
+		RequeueAfter: time.Second * 5,
+	}
+
+	o := &pxcv1.PerconaXtraDBCluster{}
+	err := r.Client.Get(context.TODO(), req.NamespacedName, o)
+	if err != nil {
+		return rr, err
+	}
+
+	if _, ok := o.Annotations["percona.com/issue-vault-token"]; !ok {
+		return rr, nil
+	}
+
+	log.Info("found annotation")
+
+	opNs, err := operatorNamespace()
+	if err != nil {
+		return rr, errors.Wrap(err, "get operator namespace")
+	}
+
+	newSecretObj := corev1.Secret{}
+	err = r.Client.Get(context.TODO(),
+		types.NamespacedName{
+			Namespace: o.Namespace,
+			Name:      o.Spec.VaultSecretName,
+		},
+		&newSecretObj,
+	)
+	if !apierrors.IsNotFound(err) {
+		return rr, err
+	}
+	if err == nil {
+		log.Info("issued secret was found, delete annotaton")
+		err = r.deleteAnnotation(o)
+		return rr, err
+	}
+
+	rootSecretName, err := rootSecretName()
+	if err != nil {
+		return rr, err
+	}
+
+	rootSecretObj := corev1.Secret{}
+	err = r.Client.Get(context.TODO(),
+		types.NamespacedName{
+			Namespace: opNs,
+			Name:      rootSecretName,
+		},
+		&rootSecretObj,
+	)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("root secret was not found")
+			return rr, nil
+		}
+		return rr, err
+	}
+
+	err = r.IssueVaultToken(rootSecretObj, o.Spec.VaultSecretName, o.Namespace)
+	if err != nil {
+		return rr, err
+	}
+
+	log.Info("token was issued, delete annotation")
+
+	return rr, r.deleteAnnotation(o)
+}
+
+func (r *PerconaXtraDBClusterReconciler) deleteAnnotation(o *pxcv1.PerconaXtraDBCluster) error {
+	return r.Client.Patch(context.Background(), o, client.RawPatch(types.JSONPatchType, []byte("[{\"op\": \"remove\", \"path\": \"/metadata/annotations/percona.com~1issue-vault-token\"}]")))
```

Comment:
percona.com~1issue-vault-token
`~1issue` typo?

Reply:
no, it's needed because `/` character in `percona.com/issue-vault-token` should be escaped
and this is how it's done

Reply:
kk then =)